File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/util/GPSTracker.java
package com.antonio.samir.meteoritelandingsspots.util;


import android.Manifest;
import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.util.Log;

import com.antonio.samir.meteoritelandingsspots.Application;


public final class GPSTracker implements LocationListener {

    // The minimum distance to change Updates in meters
    private static final long MIN_DISTANCE_CHANGE_FOR_UPDATES = 1; // 10 meters
    // The minimum time between updates in milliseconds
    private static final long MIN_TIME_BW_UPDATES = 1; // 1 minute
    private static final String TAG = GPSTracker.class.getSimpleName();
    private static boolean permissionRequested = false;
    private final Context mContext;
    private final GPSTrackerDelegate mDelegate;
    // flag for GPS status
    public boolean isGPSEnabled = false;
    // Declaring a Location Manager
    protected LocationManager locationManager;
    // flag for network status
    boolean isNetworkEnabled = false;
    // flag for GPS status
    boolean canGetLocation = false;

    private MutableLiveData<Location> liveLocation; // liveLocation

    public void stopUpdates() {
        locationManager.removeUpdates(this);
    }

    public interface GPSTrackerDelegate {
        void requestPermission();
    }

    public GPSTracker(GPSTrackerDelegate delegate) {
        mDelegate = delegate;
        liveLocation = new MutableLiveData<>();
        this.mContext = Application.getContext();
    }

    /**
     * Function to get the user's current liveLocation
     *
     * @return
     */
    public MutableLiveData<Location> getLocation() {
        try {
            startLocationService();
        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
        }
        return liveLocation;
    }

    public void startLocationService() {
        try {
            final boolean isGPSEnabled = isGPSEnabled();
            if (isGPSEnabled) {

                if (!permissionRequested && ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                    mDelegate.requestPermission();
                    permissionRequested = true;
                } else {

                    this.canGetLocation = true;

                    Location location = null;

                    if (isNetworkEnabled) {
                        liveLocation = null;

                        locationManager.requestLocationUpdates(
                                LocationManager.NETWORK_PROVIDER,
                                MIN_TIME_BW_UPDATES,
                                MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
                        Log.d(TAG, "Network");
                        if (locationManager != null) {
                            location = locationManager
                                    .getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
                            if (location != null) {
                                onLocationChanged(location);
                            }
                        }
                    }
                    // if GPS Enabled get lat/long using GPS Services
                    if (this.isGPSEnabled) {
                        location = null;
                        if (location == null) {
                            locationManager.requestLocationUpdates(
                                    LocationManager.GPS_PROVIDER,
                                    MIN_TIME_BW_UPDATES,
                                    MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
                            Log.d(TAG, "GPS Enabled");
                            if (locationManager != null) {
                                location = locationManager
                                        .getLastKnownLocation(LocationManager.GPS_PROVIDER);
                                onLocationChanged(location);
                            }
                        }
                    }
                }

            }

        } catch (Exception e) {
            Log.e(TAG, e.getMessage(), e);
        }
    }

    public boolean isGPSEnabled() {
        locationManager = (LocationManager) mContext
                .getSystemService(Context.LOCATION_SERVICE);

        // getting GPS status
        isGPSEnabled = locationManager
                .isProviderEnabled(LocationManager.GPS_PROVIDER);

        Log.v("isGPSEnabled", "=" + isGPSEnabled);

        // getting network status
        isNetworkEnabled = locationManager
                .isProviderEnabled(LocationManager.NETWORK_PROVIDER);

        Log.v("isNetworkEnabled", "=" + isNetworkEnabled);

        return false && isGPSEnabled || isNetworkEnabled;
    }


    @Override
    public void onLocationChanged(final Location location) {
        liveLocation.setValue(location);
    }

    @Override
    public void onProviderDisabled(String provider) {
    }

    @Override
    public void onProviderEnabled(String provider) {
    }

    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {
    }

}

File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/util/GPSTracker.kt
package com.antonio.samir.meteoritelandingsspots.util

import android.Manifest
import android.Manifest.permission.ACCESS_COARSE_LOCATION
import android.content.Context
import android.content.Context.LOCATION_SERVICE
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.location.Location
import android.location.LocationManager
import android.util.Log
import androidx.core.app.ActivityCompat.checkSelfPermission
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import java.util.concurrent.atomic.AtomicBoolean

@FlowPreview
@ExperimentalCoroutinesApi
class GPSTracker(private val context: Context) : GPSTrackerInterface {

    // flag for GPS status
    private var isGPSEnabled = false

    // Declaring a Location Manager
    private var locationManager = context.getSystemService(LOCATION_SERVICE) as LocationManager

    private val isLocationServiceStarted = AtomicBoolean(false)

    private var currentLocation = MutableStateFlow<Location?>(null)

    /**
     * Function to get the user's current liveLocation
     *
     * @return Location
     */
    override val location: Flow<Location?>
        get() = currentLocation

    private var currentNeedAuthorization = MutableSharedFlow<Boolean>()

    /**
     * Function to know if the authorization is required or not
     *
     * @return Boolean
     */
    override val needAuthorization: Flow<Boolean>
        get() = currentNeedAuthorization

    override fun isLocationServiceStarted(): Boolean = isLocationServiceStarted.get()

    override val isLocationAuthorized: Boolean
        get() {
            return checkSelfPermission(context, ACCESS_COARSE_LOCATION) == PERMISSION_GRANTED
        }

    override suspend fun stopUpdates() {
        isLocationServiceStarted.set(false)
//        locationManager.removeUpdates(this)
    }

    override suspend fun requestLocation() {
        if (isGPSEnabled()) {
            try {
                if (isLocationAuthorized) {
                    // if GPS Enabled get lat/long using GPS Services
                    if (isGPSEnabled) {
                        //            locationManager.requestLocationUpdates(
                        //                    LocationManager.GPS_PROVIDER,
                        //                    MIN_TIME_BW_UPDATES,
                        //                    MIN_DISTANCE_CHANGE_FOR_UPDATES.toFloat(),
                        //                    this@GPSTracker)
                        Log.d(TAG, "GPS Enabled")
                        if (checkSelfPermission(
                                context,
                                Manifest.permission.ACCESS_FINE_LOCATION
                            ) != PERMISSION_GRANTED && checkSelfPermission(
                                context,
                                ACCESS_COARSE_LOCATION
                            ) != PERMISSION_GRANTED
                        ) {
                            return
                        }
                        val lastKnownLocation =
                            locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER)
                        Log.i(TAG, "Location received $lastKnownLocation")
                        currentLocation.emit(lastKnownLocation)
                    } else {
                        currentLocation.emit(null)
                    }
                }
                currentNeedAuthorization.emit(!isLocationAuthorized)
            } catch (e: Exception) {
                Log.e(TAG, e.message, e)
            }
        }
    }

    override fun isGPSEnabled(): Boolean {

        // getting GPS status
        isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)

        return isGPSEnabled
    }


    companion object {

        private val TAG = GPSTracker::class.java.simpleName
    }

}